// s3-bucket/index.js
var knox = require('knox')
, Resource = require('deployd/lib/resource')
, httpUtil = require('deployd/lib/util/http')
, formidable = require('formidable')
, fs = require('fs')
, util = require('util')
, path = require('path');


function S3Bucket(name, options) {
  Resource.apply(this, arguments);
  if (this.config.key && this.config.secret && this.config.bucket) {
    this.client = knox.createClient({
        key: this.config.key
      , secret: this.config.secret
      , bucket: this.config.bucket
    });
  }
}
util.inherits(S3Bucket, Resource);
module.exports = S3Bucket;
S3Bucket.label = "S3 Bucket";

S3Bucket.prototype.clientGeneration = true;

S3Bucket.events = ["upload", "get", "delete"];
S3Bucket.basicDashboard = {
  settings: [{
      name: 'bucket'
    , type: 'string'
  }, {
      name: 'key'
    , type: 'string'
  }, {
      name: 'secret'
    , type: 'string'
  }]
};

S3Bucket.prototype.handle = function (ctx, next) {
  var req = ctx.req
    , bucket = this
    , domain = {url: ctx.url};

  if (!this.client) return ctx.done("Missing S3 configuration!");
    
  if (req.method === "POST" && !req.internal && req.headers['content-type'].indexOf('multipart/form-data') === 0) {
    var form = new formidable.IncomingForm();
    var remaining = 0;
    var files = [];
    var error;

    function uploadedFile(err) {
      if (err) {
        error = err;
        return ctx.done(err);
      } else if (!err) {
        remaining--;
        if (remaining <= 0) {
           if (req.headers.referer) {
             httpUtil.redirect(ctx.res, req.headers.referer || '/');
          } else {
            ctx.done(null, files);
          }
        }
      }
    };

    form.parse(req)
      .on('file', function(name, file) {
        remaining++;
          
        if (bucket.events.upload) {
          bucket.events.upload.run(ctx, {url: ctx.url, fileSize: file.size, fileName: file.filename}, function(err, res) {
            if (err) return uploadedFile(err);
            bucket.uploadFile(file.filename, file.size, file.mime, fs.createReadStream(file.path), uploadedFile, ctx);
          });
        } else {
          bucket.uploadFile(file.filename, file.size, file.mime, fs.createReadStream(file.path), uploadedFile, ctx);
        }
      })

      .on('error', function(err) {
        ctx.done(err);
        error = err;
      });
    req.resume();
    return;
  }

  if (req.method === "POST" || req.method === "PUT") {
      
    domain.fileSize = ctx.req.headers['content-length'];
    domain.fileName = path.basename(ctx.url);
    
    ctx.dpd.files.post({fileName: domain.fileName,fileSize: domain.fileSize}
                         ,function(res, err) {
                         if (err) return ctx.done(err);
                         bucket.upload(res, ctx, next);
                         });
    
  } else if (req.method === "GET") {
    if (ctx.res.internal) return next(); 

    if (this.events.get) {
      this.events.get.run(ctx, domain, function(err) {
        if (err) return ctx.done(err);
        bucket.get(ctx, next);
      });
    } else {
      this.get(ctx, next);
    }

  } else if (req.method === "DELETE") {
    
    if (this.events['delete']) {
      this.events['delete'].run(ctx, domain, function(err) {
        if (err) return ctx.done(err);
        bucket.del(ctx, next);
      }); 
    } else {
      this.del(ctx, next);
    }
  } else {
    next();
  }
}

S3Bucket.prototype.uploadFile = function(filename, filesize, mime, stream, fn, ctx) {
  var bucket = this;

  var headers = {
      'Content-Length': filesize
    , 'Content-Type': mime
  };

  this.client.putStream(stream, filename, headers, function(err, res) {
    if (err) return ctx.done(err);
    if (res.statusCode !== 200) {
      bucket.readStream(res, function(err, message) {
        fn(err || message);
      });
    } else {
      fn();
    }
  });
};

S3Bucket.prototype.upload = function(result, ctx, next) {
  var bucket = this
    , req = ctx.req
    , fileId = result.id;
    
  var headers = {
      'Content-Length': req.headers['content-length']
    , 'Content-Type': req.headers['content-type']
  };

  this.client.putStream(req, fileId, headers, function(err, res) {    
    if (err) return ctx.done(err);
    if (res.statusCode !== 200) {
      bucket.readStream(res, function(err, message) {
        ctx.done(err || message);
      });
    } else {              
      ctx.done(result);
    }
  });
  req.resume();
}

S3Bucket.prototype.get = function(ctx, next) {
  var bucket = this;
  this.client.get(ctx.url).on('response', function(res) {
    if (res.statusCode === 200) {
      res.pipe(ctx.res); 
    } else {
      bucket.readStream(res, function(err, message) {
        ctx.done(err || message);
      });
    }
  }).on('error', function(err) {
    ctx.done(err);
  }).end();
}

    
S3Bucket.prototype.del = function(ctx, next) {
  var bucket = this;

  this.client.deleteFile(ctx.url, function(err, res) {
    if (err) ctx.done(err);
    if (res.statusCode !== 200) {
      bucket.readStream(res, function(err, message) {
        ctx.done(err || message);
      });
    } else {
      ctx.done();
    }
  });
};

S3Bucket.prototype.readStream = function(stream, fn) {
  var buffer = '';
  stream.on('data', function(data) {
    buffer += data;
  }).on('end', function() {
    fn(null, buffer);
  }).on('error', function(err) {
    fn(err);
  });
};
